#369 - Work Package Form Validation

z-roth opened the pull request:
This ticket is not complete - opening for review.

**Completed**
-Duration and progress fields can already only contain numbers as they are number inputs. I set the minimum value to 0 as they will not be negative.
-WBS# and Expected Progress are read-only, and have no validation or editability.
-None of the bullet list items (Expected Activities, Deliverables) can have an empty textbox (unless it is the add new bullet textbox)

**Uncompleted**
- [x]  Validation for Dependencies. Stuck on trying to figure out an implementation for it - will use try and catch with ValidateWBS, but working on creation of an on-screen validation error message.
- [x]  Create function to reformat startDate to "YYYY-MM-DD" so it can be used as a default value for the date.

**Notes/Questions**
- Should we remove the capability of the CR ID being picked by a user? Adds an extra step to validation - having to make the async call to check if it is taken. Instead we can make it randomly generated when submitted.
- Dependencies can only be added once at a time with current implementation? Does + button add a new dependency input?

Resolves #369

jpwaves commented:
I've generally handled CR ID validation by just making sure the number inputted could potentially be a CR ID and letting the backend handle verifying that the number given is a valid CR ID, which is not the best way to do validation but it seemed the fastest way to get it done. As far as dependencies go, adding a new dependency should create a new input box for it and dependencies I believe are added one at a time. If you're using the `EditableTextInputList` component for getting the list of dependencies in the form, all the dependencies would be stored in a state variable with the dependencies represented as a list of strings.

jamescd18 commented on src/components/shared/editable-text-input-list/editable-text-input-list.tsx:
```diff
@@ -3,37 +3,65 @@
+  if (!readOnly) {
+    listPrepared = [...listPrepared, addButton];
+  }
+  let builtList = <ul>{listPrepared}</ul>;
+  if (ordered) {
+    builtList = <ol>{listPrepared}</ol>;
+  }
```
I don't think we need nor want to have HTML list tags in this stuff. If it isn't in there for the project edit form then let's not add it for this.

It looks like unnecessary indentation to me
<img width="818" alt="Screen Shot 2022-03-14 at 11 06 22 PM" src="https://user-images.githubusercontent.com/51571627/158297996-2eea0174-4635-4bf4-93c1-d0dbf287ab3c.png">

jpwaves replied:
I think it's fine if you want to turn this into a list HTML element, but I think a better way to go about it would be to replace the `div` element in the `map` with `li` and then wrap the output of the `map` in an `ul` so that it's something like `<ul>{mappedItems}</ul>`. Only thing with this is that you need to make sure to remove the default stylings from the list HTML elements (mainly `ul`, not sure if `li` also has default styling that needs to be reset)

jamescd18 reviewed (changes requested):
This is amazing work! There's a few different things I think need to change before it's ready to go, but it looks to be most of the way there. Also, be sure to use the new PR template before pinging for a second review. Awesome job!

jamescd18 commented:
## Changes

_Explanation of changes goes here_

## Notes

_Any other notes go here_

## Test Cases

- Case A
- Edge case
- ...

## Screenshots (if applicable)

_Any relevant screenshots go here_

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/blob/main/docs/ContributorGuide.md) and reach out to your squad if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors
- [ ] No newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (if applicable)
- [ ] Remove any not-applicable sections
- [ ] Assign the PR to yourself
- [ ] PR is linked to the ticket
- [ ] No `package-lock.json` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack

Closes # (issue #)

z-roth commented on src/components/projects/wbs-details/work-package-container/work-package-container.tsx:
```diff
@@ -30,24 +35,105 @@ export interface EditModeProps {
+  const [expectedActivities, setExpectedActivities] = useState(['']);
+  const [deliverables, setDeliverables] = useState(['']);
+
+  // On render, data is undefined. This causes issues if trying to set the state to fields from it. Bug?
```
Should I do anything else about this?

jpwaves replied:
I think how I resolved this is to make a wrapper component that gets the data separately and then passes that data to the view component (I presume that is this). That way, the view component will be guaranteed to always have data that is defined such that you show the loading indicator while you wait for the data, and then once the data is retrieved, you render the form with the data at the same time.

z-roth replied:
Is it alright if I keep it like this? It works, but I'm also a bit confused about how I could implement such a wrapper. Should I create a component up or down the tree?

jpwaves replied:
If you went with a wrapper component, you'd make a component up the tree. It would do all the data processing and contain all the business logic like handler functions and things of that sort, and then pass that all down to another component (which I believe is what this is) that acts as a view. There are quite a few examples of this pattern throughout the codebase -- a simple example of this would be the review change request form if you want to check that out.

z-roth commented on src/components/projects/wbs-details/work-package-container/work-package-details/work-package-details.test.tsx:
```diff
@@ -60,8 +57,8 @@ describe('Rendering Work Package Details Component', () => {
     ).toBeInTheDocument();
-    const progresses = screen.getAllByText(`${percentPipe(wp.progress)}`); // progress and expectedProgress should be equal and return 2 results
-    expect(progresses.length).toBe(2);
+    const progresses = screen.getAllByText(`${percentPipe(wp.progress)}`); // expected progress has been removed, so only 1 progress at the moment.
```
Honestly, not really sure. I saw that this test was counting on Progress and Expected Progress being rendered, but it looks like ExpectedProgress doesn't exist anymore with one of the more recent pulls. This was causing tests to fail.

jpwaves commented on src/components/shared/editable-text-input-list/editable-text-input-list.tsx:
```diff
@@ -3,37 +3,65 @@
   remove,
   update
 }) => {
+  let listPrepared = items.map((item, index: number) =>
+    !readOnly ? (
+      <div key={index} className={'mb-2'}>
+        <InputGroup>
+          <Form.Control
+            type="text"
+            value={item.toString()}
+            onChange={(e) => {
+              update(index, e.target.value);
+            }}
+          />
+          <Button type="button" variant="danger" onClick={() => remove(index)}>
+            X
+          </Button>
+        </InputGroup>
+      </div>
+    ) : (
+      <li key={index}>{item}</li>
+    )
+  );
```
`Form.Control` has a prop to make it readonly, so there's no need for the ternary operator. Also, if we're making the list readOnly, we'd probably want a ternary operator on the remove button instead (if `readOnly`, don't render anything aka `<></>`; otherwise, render the remove button)

z-roth replied:
True, but the readOnly style looks like a greyed-out textbox. Do we want this kind of styling? From a UI/UX standpoint I feel like it would be a little confusing.

jpwaves replied:
Not really sure tbh, I'd ask the PMs about that in the product-request or product channel. Although, this is just MVP so I've been operating under the impression that UI doesn't need to be perfect and so I've let things like this just slide.

z-roth replied:
Sent a message in #product-request.

anthonybernardi replied:
personally I prefer the <li> and it doesn't really add any complexity i think

jpwaves commented on src/components/shared/editable-text-input-list/editable-text-input-list.tsx:
```diff
@@ -3,37 +3,65 @@
+  title,
   items,
+  readOnly,
+  ordered,
```
I'm not sure how `ordered` would work in this type of list since the list items are text inputs and those can be changing at any time, so depending on what is in each text input at a given time, the order may be different. How were you envisioning `ordered` to work?

jamescd18 replied:
I think I mentioned the whole list tag stuff should be removed

jpwaves replied:
Yeah, I saw that - semantically speaking, it would make sense to make this use a list tag as opposed to `div`s. Only caveat would be to remove all the default stylings on `li`, `ul`, and `ol` tags and then apply the styling that is currently on the `div`

z-roth replied:
When I implemented this, I was looking at the previous implementation (the BulletList component) to see how I could port it over to the complete EditableTextInputList. Should I remove this or keep it in?

jpwaves replied:
If it's easier for you to keep it, I think it's fine to keep it. We can always make a ticket (and I hope we do) to refine EditableTextInputList in the future

jpwaves commented on src/components/shared/bullet-list/bullet-list.tsx:
```diff
@@ -6,30 +6,71 @@
+  let listPrepared = bullets.map((bullet, idx) =>
     editMode && !readOnly ? (
-      <InputGroup>
-        <Form.Control type="text" defaultValue={bullet.props.children} placeholder={bullet.props.children} key={idx} />
-        <Button variant="danger">X</Button>
+      <InputGroup aria-required>
+        <Form.Control
+          required
+          type="text"
+          defaultValue={bullet.props.children}
+          placeholder={bullet.props.children}
+          key={idx}
+          onChange={(e) => handleChange(`${fieldName}${idx}`, e.target.value)}
+        />
+        <Button variant="danger" key={idx} onClick={handleDelete}>
+          X
+        </Button>
       </InputGroup>
     ) : (
       <li key={idx}>{bullet}</li>
```
Same comment and recommended edits as with the `EditableTextInputList`, `readOnly ` is a prop on `Form.Control` that you can use and we can probably take out the remove button rendering when it is read only.

jpwaves commented on src/components/projects/wbs-details/work-package-container/work-package-details/work-package-details.tsx:
```diff
@@ -9,7 +9,7 @@ import PageBlock from '../../../../shared/page-block/page-block';
 import { Col, Container, Row, Form } from 'react-bootstrap';
```
idk why I can't select line 7, but Github is showing a warning that `percentPipe` isn't ever used.

jpwaves commented on src/components/projects/wbs-details/work-package-container/dependencies-list/dependencies-list.tsx:
```diff
@@ -7,22 +7,57 @@ import HorizontalList from '../../../../shared/horizontal-list/horizontal-list';
   );
-  const items = dependencies.map((ele) => <Dependency wbsNumber={ele} />);
+
+  function handleDelete(dependencyToDelete: WbsNumber) {
+    const index = dependenciesState.indexOf(dependencyToDelete);
+    if (index !== -1) {
+      const half1 = dependenciesState.slice(0, index);
+      const half2 = dependenciesState.slice(index + 1);
+      setDependenciesState(half1.concat(half2));
+    }
+  }
+
+  function handleAdd() {
+    try {
+      setDependenciesState([...dependenciesState, validateWBS(unvalidatedDependency)]);
+    } catch (error: any) {
+      alert(error.message);
+    }
+    setUnvalidatedDependency('');
+  }
```
@jamescd18 correct me if I'm wrong, but I think we generally try to use ES6 arrow function expressions over function declarations

jpwaves commented on src/components/shared/editable-text-input-list/editable-text-input-list.tsx:
```diff
@@ -3,37 +3,65 @@
-      }
-      <Button type='button' variant='success' onClick={() => add('')}>+ Add New Bullet</Button>
-    </>
+    <PageBlock title={title} headerRight={<></>} body={<Form.Group>{builtList}</Form.Group>} />
```
I don't think it's a good idea to wrap the content of the `EditableTextInputList` in a `PageBlock`, reason being that it would make the styling very weird if we wanted to have multiple `EditableTextInputList`s in a single `PageBlock` or some other component. It does mean that there is more work to do in other higher-level components, but that also means this component can be easily swapped out in the future if we decide to make a new implementation of `EditableTextInputList` (which I hope we do because I made this very quickly so there are quite a few issues with it) or decide we don't want to use it anymore

z-roth replied:
Thoughts on adding a PageBlock prop? So it can be rendered in it's own pageblock or without it

jpwaves replied:
If you add a PageBlock prop, that would be tightly coupling this component to the PageBlock component, which is what I'm trying to avoid. Even if we end up continuing to use PageBlock's for our UI, I think it's still bad design to have components reliant on each other in this manner, especially with how versatile the use cases for an EditableTextInputList can be.

jpwaves reviewed (changes requested):
Just my initial thoughts on looking at the code, don't have the time tonight to manually test but I'll do that tomorrow

z-roth commented:
## Changes

- Implemented date formatter for default date value for EditableDetail
- Implemented add, delete, and validate functionalities for Dependency component.
- Added min and max fields to editable details of number type
- Implemented EditableTextInputList for Work Package Edit Form, added readOnly functionality
- Converted handleAdd and handleDelete functions to arrow functions
- Created view component for work package container

## Notes

## Test Cases

- No new testing due to no new major UI components.

## Screenshots (if applicable)

_Any relevant screenshots go here_

## To Do

- [ ] Fix failing tests

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/blob/main/docs/ContributorGuide.md) and reach out to your squad if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors
- [ ] No newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (if applicable)
- [ ] Remove any not-applicable sections
- [ ] Assign the PR to yourself
- [ ] PR is linked to the ticket
- [ ] No `package-lock.json` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack

Closes # (issue #)

jamescd18 reviewed (changes requested):
Not sure exactly which test but it looks like the test is failing because the WBS # is different. one is 1.12.0 and the other is 1.1.1. So that's gotta be fixed one way or the other to avoid that situation.

<img width="1199" alt="Screen Shot 2022-03-24 at 9 30 29 PM" src="https://user-images.githubusercontent.com/51571627/160036967-0ec03350-c16b-4b58-8c7f-6d3aaf3efd47.png">

z-roth commented:
@jpwaves @jamescd18 could one of you guys check out these merge conflicts? I'm a bit confused on how to fix this. @anthonybernardi and I took a look into it today at our meeting, but nothing worked.

acd124 commented on src/components/projects/wbs-details/work-package-container/work-package-container.module.css:
```diff
@@ -1,5 +1,4 @@
  * See the LICENSE file in the repository root folder for details.
- */
-
+ */
```
This is also apparently a merge conflict, I would recommend letting prettier run on the file, so hopefully it would end up not changed

acd124 commented:
`src/components/projects/wbs-details/work-package-container/dependencies-list/dependency/dependency.test.tsx` this file, and the other file in the same folder it said something about the file having changes in main but being deleted in this branch, I'm not sure how best to fix it other than merging main and making the delete again, not sure what actually happened for those two files though

jamescd18 reviewed (changes requested):
Can you please change the project lead and project manager fields so that they're instead dropdowns populated with the list of users? The logic should be available to understand and copy from the project edit form.

jamescd18 reviewed (approved):
LGTM! Soooo exciting!

z-roth commented:
Woot!!! 🥳🍾
